Add authentication design proposal for PetMatch

BDubDesigns opened the pull request:
## Description

Brief summary of changes made in this pull request.
- Created an auth_proposal.md doc in the /doc/ folder with a proposal for how our authentication flow could work. I am more than open to any changes, this is just what I came up with after a few hours of research :)
- For task #8 

## Check list

- [X] Have you rebased the branch on top of the latest change in main?
- [x] Is your code getting reviewed?
- [x] Has you squashed your code after getting approved?

qvd808 commented:
> > I think overall it is a really good first sketch. I would say you can add more of block diagram to represent the logic of how the frontend going to interact and request authentication from the backend. It seems like the idea is there but having a diagram would flesh it out more. I also want to go more details into how we are going to implementation. Like the function that will be created on the backend and frontend and their purpose
> 
> Thank you very much for that, I will take another pass at it going more in depth on implementation. You're right, it's very surface level, I will flesh out the actual logic and functions. Should I cancel this PR and resubmit another one when I'm done, or what is best practice? Thanks again!

You can take a look at this video. I want something similar to this diagram where we can understand the flow of the operation. How the user is going to request the authentication from frontend and what's going to be happened and through which route of API: https://www.youtube.com/watch?v=WzcKgo7BQC0

adamocamb commented:
This is awesome! I had noticed that in Gio's frontend doc she mentioned that we could optionally manually verify a shelter so that not just anyone can make an account and list pets. I think something like that would be compatible with this flow if we wanted to do that, is that correct?

BDubDesigns commented:
> This is awesome! I had noticed that in Gio's frontend doc she mentioned that we could optionally manually verify a shelter so that not just anyone can make an account and list pets. I think something like that would be compatible with this flow if we wanted to do that, is that correct?

That should be no problem, yeah! Instead of the shelters creating their accounts, we could have an admin role for us to manage the site and add shelters. Maybe anyone can apply as a shelter, and then we have an admin dashboard to approve or deny their application to become a shelter on our site? That way we can collect the information and hash the password like normal on a registration page, they just wouldn't be approved until one of us manually approves them while logged into an account with admin role. Does that sound like the way we want to go? Or what should the flow look like for getting their username and hashed pass and email from the shelters?

That's a good idea not letting just anyone create a shelter account, and I can't believe I didn't have an admin role for us already written up smh 🤦

adamocamb commented:
> Instead of the shelters creating their accounts, we could have an admin role for us to manage the site and add shelters. Maybe anyone can apply as a shelter, and then we have an admin dashboard to approve or deny their application to become a shelter on our site? That way we can collect the information and hash the password like normal on a registration page, they just wouldn't be approved until one of us manually approves them while logged into an account with admin role. Does that sound like the way we want to go?

Yeah I like that, that sounds good

qvd808 commented:
Great job @BDubDesigns, the diagram is very clear. I think Adam and I can create task based on your workflow. @uhohgio and @JackSawyerWATX can you verified the diagram? Just see if there is anything else missing. Once you guys give the check, then I will approve and merge this PR

qvd808 commented:
You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase

BDubDesigns commented:
> You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase

Crap, I'm sorry, I thought that only applied to merging to main, sorry! Thanks for pointing it out! Should I reset and rebase auth_proposal off main now and force-push to clean it up? Or what should I do going forward to fix this?

qvd808 commented:
> > You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase
> 
> Crap, I'm sorry, I thought that only applied to merging to main, sorry! Thanks for pointing it out! Should I reset and rebase auth_proposal off main now and force-push to clean it up? Or what should I do going forward to fix this?

Yes please, probally reset the merge commit and rebase on top of main branch is good.

BDubDesigns commented:
> > > You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase
> > 
> > 
> > Crap, I'm sorry, I thought that only applied to merging to main, sorry! Thanks for pointing it out! Should I reset and rebase auth_proposal off main now and force-push to clean it up? Or what should I do going forward to fix this?
> 
> Yes please, probally reset the merge commit and rebase on top of main branch is good.

Will do on lunch break! Just started work or id do it now :) thanks for the help!

BDubDesigns commented:
> > > You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase
> > 
> > 
> > Crap, I'm sorry, I thought that only applied to merging to main, sorry! Thanks for pointing it out! Should I reset and rebase auth_proposal off main now and force-push to clean it up? Or what should I do going forward to fix this?
> 
> Yes please, probally reset the merge commit and rebase on top of main branch is good.

Ok, I thiiiink I got it all cleaned up and fixed. After I reset the previous commit I realized that I had merged the main branch into auth_proposal earlier as well, so I went all the way back to my first clean commit, and then re added the 2 files I worked on after rebasing. So I think I have it right, but if its cleaner / better to just delete this branch and then create a new one with those 2 files, I'd be happy to do that instead too. I just figured I'd try to clean it up first for the practice / experience. Thanks!

qvd808 commented:
> > > > You should rebase instead of merging the main branch @BDubDesigns. If you use the Update button, there should be a drop down indicate that you should rebase
> > > 
> > > 
> > > Crap, I'm sorry, I thought that only applied to merging to main, sorry! Thanks for pointing it out! Should I reset and rebase auth_proposal off main now and force-push to clean it up? Or what should I do going forward to fix this?
> > 
> > 
> > Yes please, probally reset the merge commit and rebase on top of main branch is good.
> 
> Ok, I thiiiink I got it all cleaned up and fixed. After I reset the previous commit I realized that I had merged the main branch into auth_proposal earlier as well, so I went all the way back to my first clean commit, and then re added the 2 files I worked on after rebasing. So I think I have it right, but if its cleaner / better to just delete this branch and then create a new one with those 2 files, I'd be happy to do that instead too. I just figured I'd try to clean it up first for the practice / experience. Thanks!

I think what you did seems pretty good. Nice job on the clean up

BDubDesigns commented:
Alright, if I've done everything correctly, it is now rebased and I squashed my commits into a single commit (first time doing it, so hopefully I did it right lol) so I thiiiink its ready to merge, if its indeed signed off by a reviewer.

uhohgio commented:
It looks awesome! The only thing that needs clarity is the process by which shelters can secure a shelter username and shelter password. As someone who just popped in, it's unclear as to how that process happens (in-app or otherwise). If the process happens off the app, that is fine, but you should explicitly mention that. If the process is in-app, however, you should mention the application process by which a shelter would go about applying for a shelter profile. Personally, considering the scope of our project and the speed at which we are progressing, I think we should take it off the app IF you haven't already made a decision. We can do a simple "contact us if you want to apply as a shelter" type of thing, but then make sure that the documentation says that this is how shelter accounts are made. That way, anyone who isn't caught up isn't confused!

uhohgio commented:
Positive Notes: YES I LOVE THE BCRYPT! Password security is something that always gets overlooked with these kinds of things (I am guilty). I also think the diagrams and your explanations are fantastic! I just got here, and everything is clear and thorough, as all documentation should strive to be. :) Great work.

BDubDesigns commented:
> It looks awesome! The only thing that needs clarity is the process by which shelters can secure a shelter username and shelter password. As someone who just popped in, it's unclear as to how that process happens (in-app or otherwise). If the process happens off the app, that is fine, but you should explicitly mention that. If the process is in-app, however, you should mention the application process by which a shelter would go about applying for a shelter profile. Personally, considering the scope of our project and the speed at which we are progressing, I think we should take it off the app IF you haven't already made a decision. We can do a simple "contact us if you want to apply as a shelter" type of thing, but then make sure that the documentation says that this is how shelter accounts are made. That way, anyone who isn't caught up isn't confused!

I think that's a good idea to just keep it off, I'll be off work shortly here at 10 and update that documentation as soon as I'm home and squash the commits :) thanks!

qvd808 commented:
> le "contact us if you want to apply as a shelter" type of thing, but then make sure that the documentation says that this is how shelter accounts are made. That way, anyone who isn't caught up isn't confus

Shouldn't shelter sign up process should be similar except that we sign up as a shelter. It is similar to other online store where you can sign up as a user or buyer

uhohgio commented:
> > > > le "contact us if you want to apply as a shelter" type of thing, but then make sure that the documentation says that this is how shelter accounts are made. That way, anyone who isn't caught up isn't confus
> > > 
> > > 
> > > Shouldn't shelter sign up process should be similar except that we sign up as a shelter. It is similar to other online store where you can sign up as a user or buyer
> > 
> > 
> > Without verifying? I thought we were only letting verified shelters post? Maybe I misunderstood.
> 
> I guess the misunderstanding come from how the verification process work? Is it something similar to Twitter where an account an be created then get verified later manually or should it be verified before the creation of the account?

Doable, and I know this isn't a real app, but if it were, this gets really messy. You have to have tight security or moderation, otherwise you could have accounts making fake pets. People misusing the app, trying to scam loyal customers. Essentially you open yourself up to liabilities. I guess we could make it so they can see what being a shelter is like without giving them any privileges, but it's just like overall an added layer of complexity to the project. Taking out the shelter signup reduces the amount of tasks we have to do, making a more polished version of our project more feasible.

qvd808 commented:
> > > > > le "contact us if you want to apply as a shelter" type of thing, but then make sure that the documentation says that this is how shelter accounts are made. That way, anyone who isn't caught up isn't confus
> > > > 
> > > > 
> > > > Shouldn't shelter sign up process should be similar except that we sign up as a shelter. It is similar to other online store where you can sign up as a user or buyer
> > > 
> > > 
> > > Without verifying? I thought we were only letting verified shelters post? Maybe I misunderstood.
> > 
> > I guess the misunderstanding come from how the verification process work? Is it something similar to Twitter where an account an be created then get verified later manually or should it be verified before the creation of the account?
> 
> Doable, and I know this isn't a real app, but if it were, this gets really messy. You have to have tight security or moderation, otherwise you could have accounts making fake pets. People misusing the app, trying to scam loyal customers. Essentially you open yourself up to liabilities. I guess we could make it so they can see what being a shelter is like without giving them any privileges, but it's just like overall an added layer of complexity to the project. Taking out the shelter signup reduces the amount of tasks we have to do, making a more polished version of our project more feasible.

Make sense then, we can alway add sign up for shelter in the future if we want. I just want to clarified which process we want to do. @BDubDesigns let make the shelter contact us to create an account

uhohgio commented:
Yeah, the need for clarification is exactly why we have documentation! Glad it was productive, and it seems like an easy add, so we can close this PR when Brandon gets back!
